### PR TITLE
Fixes #15407 - adding event queue and listener

### PR DESCRIPTION
**Pull request by jlsherrill**
This adds a new queing table and model as well
as a dynflow listening task that processes queue items.
The only current event is the import errata task, to serialize
errata import so that we can resolve unique constraint
violations once and for all.

### CODE REVIEW

**Comment by jlsherrill** [comment]
TODO:

*[ ]  write tests
*[ ] validate failure scenarios

**Comment by jlsherrill** [comment]
Still have a couple TODO's, but opening for early feedback

File: app/models/katello/event_queue.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module Katello
+  class EventQueue < Katello::Model
+    self.table_name = 'katello_event_queue'
+
+    @repository_types = {}
```
Comment (ehelms) [inline app/models/katello/event_queue.rb]: What does `repository` imply here? In our world, it can be an overloaded or confusing term.

File: app/models/katello/event_queue.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    def self.pop_event
+      first = Katello::EventQueue.order(:created_at => 'asc').limit(1)
+      events = Katello::EventQueue.where(:object_id => first.pluck(:object_id), :event_type => first.pluck(:event_type)).destroy_all
+      events.first
+    end
+
+    def self.push_event(event_type, id)
+      self.create!(:event_type => event_type, :object_id => id)
```
Comment (ehelms) [inline app/models/katello/event_queue.rb]: What happens when an event_type is specified that is not registered?

Comment (jlsherrill) [inline app/models/katello/event_queue.rb]: apparently my brain checked out when i was creating this class :)  That should be event types

File: app/models/katello/event_queue.rb (hunk 1 above)
Comment (jlsherrill) [inline app/models/katello/event_queue.rb]: apparently my brain checked out when i was creating this class :)  That should be event_types

File: app/models/katello/event_queue.rb (hunk 2 above)
Comment (jlsherrill) [inline app/models/katello/event_queue.rb]: I'll add throwing a more specific exception to make it obvious

File: app/models/katello/events/import_host_errata.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module Katello
+  module Events
+    class ImportHostErrata < Katello::Model
```
Comment (ehelms) [inline app/models/katello/events/import_host_errata.rb]: Does this need to be an active record object?

File: app/models/katello/event_queue.rb (hunk 2 above)
Comment (ehelms) [inline app/models/katello/event_queue.rb]: You could add a validator since the underlying object is an active record object.

**Comment by beav** [comment]
@jlsherrill I'm not 100% familiar with this area, this is an event queue that's kept inside the running instance and does not use amqp at all?

File: app/models/katello/event_queue.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+module Katello
+  class EventQueue < Katello::Model
```
Comment (ehelms) [inline app/models/katello/event_queue.rb]: My object oriented brain feels like this should be an Event with a class that represents the EventQueue and manages the events vs an EventQueue managing instances of event queues.

**Comment by jlsherrill** [comment]
@beav correct, this uses a simple db table to de-duplicate work and serialize its insertion.  We potentially could have used the AMQP queue, but this seemed much simpler

**Comment by beav** [comment]
@jlsherrill I agree, I think if it's an internal mechanism of Katello, there's no need to expose it via qpid.

File: app/lib/actions/katello/monitor_event_queue.rb [hunk 5]
```diff
@@ -0,0 +1,128 @@
+module Actions
```
Comment (ehelms) [inline app/lib/actions/katello/monitor_event_queue.rb]: For me, the triple, or quadrupel embedded classes makes it hard to parse what is going on and what belongs to what in this class (e.g. the `class << self` block).

File: app/lib/actions/katello/monitor_event_queue.rb [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      def plan
+        # Make sure we don't have two concurrent listening services competing
+        if already_running?
+          fail "Action #{self.class.name} is already active"
+        end
+        plan_self
+      end
+
+      def run(event = nil)
+        Rails.logger.error("FOOO #{event.to_s}")
```
Comment (ehelms) [inline app/lib/actions/katello/monitor_event_queue.rb]: `FOOOO`

Comment (jlsherrill) [inline app/lib/actions/katello/monitor_event_queue.rb]: BAAAAAAAAAARRRRRRRRRRR

File: app/models/katello/events/import_host_errata.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      EVENT_TYPE = 'import_host_errata'.freeze
+
+      def initialize(object_id)
+        @host = ::Host.find(object_id)
+      end
+
+      def run
+        if @host
```
Comment (ehelms) [inline app/models/katello/events/import_host_errata.rb]: Could do a `return unless @host` here

File: app/models/katello/events/import_host_errata.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+module Katello
+  module Events
+    class ImportHostErrata < Katello::Model
+      EVENT_TYPE = 'import_host_errata'.freeze
+
+      def initialize(object_id)
+        @host = ::Host.find(object_id)
```
Comment (ehelms) [inline app/models/katello/events/import_host_errata.rb]: I lean towards handling an invalid host ID properly here so as to produce a better error than the full traceback that would normally occur?

File: app/models/katello/event_queue.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    self.table_name = 'katello_event_queue'
+
+    @repository_types = {}
+
+    def self.pop_event
+      first = Katello::EventQueue.order(:created_at => 'asc').limit(1)
+      events = Katello::EventQueue.where(:object_id => first.pluck(:object_id), :event_type => first.pluck(:event_type)).destroy_all
```
Comment (ehelms) [inline app/models/katello/event_queue.rb]: Since this is doing a destroy when it pops them off, is this safe against listener failure and restarts?

File: db/migrate/20160613150922_add_event_queue.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class AddEventQueue < ActiveRecord::Migration
+  def change
+    create_table :katello_event_queue, :force => true do |t|
```
Comment (beav) [inline db/migrate/20160613150922_add_event_queue.rb]: for my understanding of how this works, this table should normally be empty?

File: app/lib/actions/katello/event_queue/monitor.rb [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+              ::Katello::EventQueue.event_class(event.event_type).new(event.object_id).run
+            end
+          rescue => e
+            output[:last_error] = e.message
+            error!(e)
+            raise e
+          ensure
+            ::Katello::EventQueue.clear_events(event.event_type, event.object_id, event.created_at)
+          end
+          suspend
```
Comment (ehelms) [inline app/lib/actions/katello/event_queue/monitor.rb]: Should we put this in the `begin` block since it would only get run if that is successful? Here it sorta looks like it should run always but due to the rescue and ensure that isn't the case.

File: app/lib/actions/katello/event_queue/poller_thread.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        def poll_for_events(suspended_action)
+          @thread.kill if @thread
+          @thread = Thread.new do
+            loop do
+              begin
+                until (event = ::Katello::EventQueue.peek_event).nil?
+                  suspended_action.notify_queue_item(event.event_type, event.object_id, event.created_at) if event
+                end
+
+                sleep 2
```
Comment (ehelms) [inline app/lib/actions/katello/event_queue/poller_thread.rb]: Just curious if there is any data or precedence for choosing 2?

Comment (jlsherrill) [inline app/lib/actions/katello/event_queue/poller_thread.rb]: hah,  5 seemed too long, 1 seemed too short.  I'm open to something different ;)

Comment (beav) [inline app/lib/actions/katello/event_queue/poller_thread.rb]: I would move this to a constant somewhere, just so its more obvious in case it needs to be tweaked later

File: app/services/katello/event_queue.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
+module Katello
+  class EventQueue
+    @event_types = {}
+
+    def self.clear_events(event_type, object_id, on_or_earlier_than)
+      Katello::Event.where(:in_progress => true, :object_id => object_id, :event_type => event_type).where('created_at <= ?', on_or_earlier_than).destroy_all
+    end
+
+    def self.peek_event
```
Comment (ehelms) [inline app/services/katello/event_queue.rb]: Do I read it correct that this function is setting all events that match the top of queue to in progress? If yes, I would not expect that a function named "peek" to change any date since it implies a viewing rather than modifying type of action.

Comment (jlsherrill) [inline app/services/katello/event_queue.rb]: suggestions?

Comment (ehelms) [inline app/services/katello/event_queue.rb]: `pop_event` ? I realize this does not remove it from the queue but it is this queues act of popping the top one off.

popop

Comment (jlsherrill) [inline app/services/katello/event_queue.rb]: but its not popping it off the top.  to me peek is a better term than pop, but i can see how both are incorrect.  'grab_task' ?

Comment (ehelms) [inline app/services/katello/event_queue.rb]: `next_event` ?

Comment (jlsherrill) [inline app/services/katello/event_queue.rb]: Works for me

File: db/migrate/20160613150922_add_event_queue.rb (hunk 10 above)
Comment (jlsherrill) [inline db/migrate/20160613150922_add_event_queue.rb]: correct.  event gets picked up (marked as in progress), event is executed, event is then deleted form the table.

**Comment by jlsherrill** [comment]
Addressed comments>  Changed the way some of the stuff worked after testing failure scenarios.  It appears that foreman-tasks already does not handle db restarts so that still is not handled after this.  

Filed:  http://projects.theforeman.org/issues/15430

When testing you can run: `Katello::EventQueue.push_event("import_host_errata", 1)`

to throw things onto the queue and see how it handles it.  I still have tests to write & fix

**Comment by jlsherrill** [comment]
This should now be ready for final testing/review

File: app/lib/actions/katello/event_queue/poller_thread.rb [hunk 14]
```diff
@@ -0,0 +1,45 @@
+module Actions
+  module Katello
+    module EventQueue
+      class PollerThread
+        SLEEP_INTERVAL = 2
```
Comment (beav) [inline app/lib/actions/katello/event_queue/poller_thread.rb]: snorlax

File: app/services/katello/event_queue.rb [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+  class EventQueue
+    @event_types = {}
+
+    def self.clear_events(event_type, object_id, on_or_earlier_than)
+      Katello::Event.where(:in_progress => true, :object_id => object_id, :event_type => event_type).where('created_at <= ?', on_or_earlier_than).destroy_all
+    end
+
+    def self.next_event
+      first = Katello::Event.where(:in_progress => false).order(:created_at => 'asc').limit(1)
+      events = Katello::Event.where(:in_progress => false, :object_id => first.pluck(:object_id),
```
Comment (beav) [inline app/services/katello/event_queue.rb]: When would the events query return more than one result?

File: db/migrate/20160613150922_add_event_queue.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
+class AddEventQueue < ActiveRecord::Migration
+  def change
+    create_table :katello_events, :force => true do |t|
+      t.integer :object_id, :null => false
+      t.string :event_type, :null => false
+      t.boolean :in_progress, :default => false, :null => false
+      t.timestamps
+    end
+
+    add_index :katello_events, [:object_id, :event_type, :in_progress], :name => :katello_events_oid_et_ip
```
Comment (beav) [inline db/migrate/20160613150922_add_event_queue.rb]: if the table gets really big (for example, if the queue consumer was dead for awhile), you may need an index on `created_at` to speed up finding the next event

File: app/services/katello/event_queue.rb (hunk 15 above)
Comment (jlsherrill) [inline app/services/katello/event_queue.rb]: If there are multiple events with the same object_id and event_type yes.  This could happen if for example two pulp applicability generation tasks finish within ~5 seconds of each other (or whatever the dynflow poll length is) and thus the applicability generation dynflow task sticks two on the queue.

File: db/migrate/20160613150922_add_event_queue.rb (hunk 16 above)
Comment (jlsherrill) [inline db/migrate/20160613150922_add_event_queue.rb]: good point, i'll go ahead and add that

File: app/services/katello/event_queue.rb (hunk 15 above)
Comment (beav) [inline app/services/katello/event_queue.rb]: Is there an easy way to pause the consumer? I would like to pause during testing and let the queue pile up a bit, then re-enable and confirm it drains

Comment (jlsherrill) [inline app/services/katello/event_queue.rb]: On a dev environment just:

1) shut down your rails server (this stops dynflow)
2) in a rails console run:
```
 Katello::EventQueue.push_event("import_host_errata", 1)
```

where '1' is the host id.  Run this as many times as you want!
When you're ready to start pulling things off the queue you can just start your rails server.   You can also put invalid host ids on the queue to simulate an error.

A 2nd option would be to just comment out `::Actions::Katello::EventQueue::Monitor.ensure_running(world)
` in engine.rb and restart your rails server.  That'll keep the monitor from running.

**Comment by beav** [comment]
This patch works functionality-wise. For any other PR testers: when you load the rails server, you'll need to hit a page to get the tasks to load up to drain the queue :bear: 

The only issue I saw was that when stopping rails server, the dynflow task did not go to "stopped/success", it stayed as "running". If I stopped and started my server 3x, it appeared as though I had three running consumers at once.

**Comment by jlsherrill** [comment]
@beav the shutting down issue should be resolved now

**Comment by beav** [comment]
@jlsherrill it works the same as the candlepin listener now in that respect :frog: 

APJ from me!

**Comment by jlsherrill** [comment]
Flipping to 'needs re-review' as @ehelms wanted to review it some more

**Comment by ehelms** [comment]
ACK - with usability comment

The current implementation doesn't make it easy to understand and know that a task triggered an applicability event that needs to be processed and that said event got processed. I don't want to hold up this change due to its importance but I think that my prior point is worth consideration.

**Comment by jlsherrill** [comment]
Agreed, opened http://projects.theforeman.org/issues/15547
